src/main/java/com/minecolonies/coremod/entity/pathfinding/EntityCitizenWalkToProxy.java

```diff
@@ -0,0 +1,206 @@
+package com.minecolonies.coremod.entity.pathfinding;
+
+import com.minecolonies.api.entity.ai.pathfinding.AbstractWalkToProxy;
+import com.minecolonies.api.util.BlockPosUtil;
+import com.minecolonies.api.util.Vec2i;
+import com.minecolonies.coremod.colony.buildings.AbstractBuildingWorker;
+import com.minecolonies.coremod.colony.buildings.BuildingMiner;
+import com.minecolonies.coremod.colony.jobs.JobBuilder;
+import com.minecolonies.coremod.colony.jobs.JobMiner;
+import com.minecolonies.coremod.entity.EntityCitizen;
+import com.minecolonies.coremod.entity.ai.citizen.miner.Level;
+import com.minecolonies.coremod.util.WorkerUtil;
+import net.minecraft.entity.EntityLiving;
+import net.minecraft.util.math.BlockPos;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+
+public class EntityCitizenWalkToProxy extends AbstractWalkToProxy
+{
+    /**
+     * Lead the miner to the other side of the shaft.
+     */
+    private static final int OTHER_SIDE_OF_SHAFT = 6;
+
+    /**
+     * The worker entity associated with the proxy.
+     */
+    private final EntityCitizen citizen;
+
+    /**
+     * Creates a walkToProxy for a certain worker.
+     * @param entity the citizen entity.
+     */
+    public EntityCitizenWalkToProxy(final EntityCitizen entity)
+    {
+        super(entity);
+        this.citizen = entity;
+    }
+
+    @Override
+    public Set<BlockPos> getWayPoints()
+    {
+        if (citizen.getColony() == null)
+        {
+            return Collections.emptySet();
+        }
+
+        return citizen.getColony().getWayPoints().keySet();
+    }
+
+    @Override
+    public boolean careAboutY()
+    {
+        return citizen.getColonyJob() instanceof JobBuilder;
+    }
+
+    @Override
+    public BlockPos getSpecializedProxy(final BlockPos target, final double distanceToPath)
+    {
+        final AbstractBuildingWorker building = citizen.getWorkBuilding();
+        if (citizen.getColonyJob() != null && citizen.getColonyJob() instanceof JobMiner && building instanceof BuildingMiner)
+        {
+            return getMinerProxy(target, distanceToPath, (BuildingMiner) building);
+        }
+        return null;
+    }
+
+    /**
+     * Returns a proxy point to the goal for the miner especially.
+     *
+     * @param target         the target.
+     * @param distanceToPath the total distance.
+     * @return a proxy or, if not applicable null.
+     */
+    @NotNull
+    private BlockPos getMinerProxy(final BlockPos target, final double distanceToPath, @NotNull final BuildingMiner building)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "getMinerProxy" is 21 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 42 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)